// @serverless/enterprise-plugin/dist/lib/safeguards/policies/allowed-function-names.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var vm = require('vm');

var _require = require('lodash'),
    entries = _require.entries,
    fromPairs = _require.fromPairs;
/*
 * Converts a string that looks like a tagged template literal into a RegExp.
 * Uses the vm module to safely eval the string as a tagged template literal.
 * The context parameter is the only thing the evaluated string is given access to.
 */


var templateStringToRegExp = function templateStringToRegExp(pattern, context) {
  return vm.runInNewContext(`new RegExp(\`^${pattern}$\`)`, context);
};

module.exports = function allowedFunctionNamesPolicy(policy, service, options) {
  var failed = false;
  var functions = service.declaration.functions,
      naming = service.provider.naming,
      Resources = service.compiled['cloudformation-template-update-stack.json'].Resources;
  var logicalFuncNamesToConfigFuncName = fromPairs(Object.keys(functions || {}).map(function (funcName) {
    return [naming.getLambdaLogicalId(funcName), funcName];
  }));
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = entries(Resources)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var _step$value = _slicedToArray(_step.value, 2),
          funcName = _step$value[0],
          _step$value$ = _step$value[1],
          Properties = _step$value$.Properties,
          Type = _step$value$.Type;

      if (Type !== 'AWS::Lambda::Function') {
        continue;
      }

      var templateContext = {
        SERVICE: service.declaration.serviceObject.name,
        STAGE: service.provider.getStage(),
        FUNCTION: logicalFuncNamesToConfigFuncName[funcName]
      };
      var regexp = templateStringToRegExp(options, templateContext);

      if (!Properties.FunctionName.match(regexp)) {
        failed = true;
        policy.fail(`Function "${logicalFuncNamesToConfigFuncName[funcName]}" doesn't match RegExp ${regexp}.`);
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator.return != null) {
        _iterator.return();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  if (!failed) {
    policy.approve();
  }
};

module.exports.docs = 'https://git.io/fjfIq';
